benchmarks/nemo/nemolite2d/manual_versions/psykal_c_opencl/time_step.c

```diff
@@ -0,0 +1,693 @@
+
+    // Boundary conditions bc_flather_u kernel (whole domain but top x boundary)
+    size_t flatheru_offset[2] = {(size_t)internal_xstart - 1, (size_t)internal_ystart - 1};
+    size_t flatheru_size[2] = {(size_t)computed_xstop, (size_t)internal_ystop + 1}; 
+    ret = clEnqueueNDRangeKernel(command_queue[0], clkernel[K_BC_FLATHER_U], 2,
+            flatheru_offset, flatheru_size, local_size, 0, NULL,
+			&(clkernevt[K_BC_FLATHER_U]));
+    check_status("clEnqueueNDRangeKernel(bc-flather-u)", ret);
+
+    // Boundary conditions bc_flather_v kernel (whole domain but top y boundary)
+    // TODO #48. flather_v kernel may have a race condition.
+    size_t flatherv_offset[2] = {(size_t)internal_xstart - 1, (size_t)internal_ystart - 1};
+    //size_t flatherv_size[2] = {(size_t)internal_xstop + 1, (size_t)internal_ystop}; 
+    size_t flatherv_size[2] = {(size_t)computed_xstop, (size_t)internal_ystop}; 
+    ret = clEnqueueNDRangeKernel(command_queue[0], clkernel[K_BC_FLATHER_V], 2,
+            flatherv_offset, flatherv_size, local_size, 0, NULL,
+			&(clkernevt[K_BC_FLATHER_V]));
+    check_status("clEnqueueNDRangeKernel(bc-flather-v)", ret);
+
+    // Copy 'next' fields to 'current' fields (whole domain)
+    ret = clEnqueueCopyBuffer(command_queue[0], ua_device, un_device, 0, 0,
+			      buff_size,0, NULL, NULL);
+    check_status("clEnqueueCopyBuffer", ret);
+    ret = clEnqueueCopyBuffer(command_queue[0], va_device, vn_device, 0, 0,
+			      buff_size,0, NULL, NULL);
+    check_status("clEnqueueCopyBuffer", ret);
+    ret = clEnqueueCopyBuffer(command_queue[0], ssha_device, sshn_device, 0, 0,
+			      buff_size,0, NULL, NULL);
+    check_status("clEnqueueCopyBuffer", ret);
+
+    // Time update kernel (internal domain u points)
+    size_t nextu_offset[2] = {(size_t)internal_xstart, (size_t)internal_ystart};
+    //size_t nextu_size[2] = {(size_t)internal_xstop - 1, (size_t)internal_ystop}; 
+    size_t nextu_size[2] = {(size_t)computed_xstop, (size_t)internal_ystop}; 
+    ret = clEnqueueNDRangeKernel(command_queue[0], clkernel[K_NEXT_SSH_U], 2,
+            nextu_offset, nextu_size, local_size, 0, NULL,
+			&(clkernevt[K_NEXT_SSH_U]));
+    check_status("clEnqueueNDRangeKernel(next-sshu)", ret);
+
+    // Time update kernel (internal domain v points)
+    size_t nextv_offset[2] = {(size_t)internal_xstart, (size_t)internal_ystart};
+    size_t nextv_size[2] = {(size_t)computed_xstop, (size_t)internal_ystop - 1};
+    ret = clEnqueueNDRangeKernel(command_queue[0], clkernel[K_NEXT_SSH_V], 2,
+            nextv_offset, nextv_size, local_size, 0, NULL,
+			&(clkernevt[K_NEXT_SSH_V]));
+    check_status("clEnqueueNDRangeKernel(next-sshv)", ret);
+
+    // Wait for the completion of all kernels  -- wait below assumes they are in-order
+    ret = clWaitForEvents(1, &(clkernevt[K_NEXT_SSH_V]));
+    check_status("clWaitForEvents", ret);
+
+#ifdef USE_TIMER
+    TimerStop();
+    TimerStart("Copy back");
+#endif
+
+    // Read the data back
```

Comment:
This bit should only be done if we're doing IO or at the end of the run but presumably it was like this before?

Reply:
Yes, the problem is that I don't know that inside the invoke now. I want to eventually reuse the dl_esm_inf get_data() infrastructure to do this copy backs (maybe giving dl_esm_inf a callback to know how to copy a field back regardless of the programming model?), but I want to do this with Kokkos in mind as well, so I propose doing it in a latter PR.

Also in terms of performance it is not much (the important part was the set-up and copy-in done just once):
```
NVIDIA 1080ti (single iteration)
OpenCLSetUp	0.5089
Continuity	0.0020
Momentum	0.0318
Remaining	0.0051
Copy back	0.0042
```

Reply:
That's surprising - I would have expected it to be quite costly but either way, if you're going to address it in the future then that's fine.